src/Proto3/Wire/Decode.hs

```diff
@@ -145,62 +113,96 @@ data ParsedField = VarintField Word64
-  where
-    kvs1 = map (\(k, v) -> (k, Data.Sequence.singleton v)) kvs0
-    hashMap = Data.HashMap.Strict.fromListWith (flip (<>)) kvs1
+toMap :: [(FieldNumber, v)] -> M.IntMap [v]
+toMap kvs0 = M.fromListWith (<>) . map (fmap (:[])) . map (first (fromIntegral . getFieldNumber)) $ kvs0
+
+decodeWire :: B.ByteString -> Either String [(FieldNumber, ParsedField)]
+decodeWire bstr = drloop bstr []
+ where
+   drloop !bs xs | B.null bs = Right $ reverse xs
+   drloop !bs xs | otherwise = do
+      (w, rest) <- takeVI' bs
+      wt <- gwireType $ fromIntegral (w .&. 7)
+      let fn = w `shiftR` 3
+      (res, rest2) <- takeWT wt rest
+      drloop rest2 ((FieldNumber fn,res):xs)
+
+
+eitherUncons :: B.ByteString -> Either String (Word8, B.ByteString)
+eitherUncons = maybe (Left "failed to parse varint128") Right . B.uncons
+
+
+takeVI' :: B.ByteString -> Either String (Word64, B.ByteString)
+takeVI' !bs =
+  case B.uncons bs of
+     Nothing -> Right (0, B.empty)
+     Just (w1, r1) -> do
+       if w1 < 128 then return (fromIntegral w1, r1) else do
+        let val1 = fromIntegral (w1 .&. 0x7F)
+
+        (w2,r2) <- eitherUncons r1
+        if w2 < 128 then return (val1 .|. (fromIntegral w2 `shiftL` 7), r2) else do
+         let val2 = (val1 .|. (fromIntegral (w2 .&. 0x7F) `shiftL` 7))
+
+         (w3,r3) <- eitherUncons r2
+         if w3 < 128 then return (val2 .|. (fromIntegral w3 `shiftL` 14), r3) else do
+          let val3 = (val2 .|. (fromIntegral (w3 .&. 0x7F) `shiftL` 14))
+
+          (w4,r4) <- eitherUncons r3
+          if w4 < 128 then return (val3 .|. (fromIntegral w4 `shiftL` 21), r4) else do
+           let val4 = (val3 .|. (fromIntegral (w4 .&. 0x7F) `shiftL` 21))
+
+           (w5,r5) <- eitherUncons r4
+           if w5 < 128 then return (val4 .|. (fromIntegral w5 `shiftL` 28), r5) else do
+            let val5 = (val4 .|. (fromIntegral (w5 .&. 0x7F) `shiftL` 28))
+
+            (w6,r6) <- eitherUncons r5
+            if w6 < 128 then return (val5 .|. (fromIntegral w6 `shiftL` 35), r6) else do
+             let val6 = (val5 .|. (fromIntegral (w6 .&. 0x7F) `shiftL` 35))
+
+             (w7,r7) <- eitherUncons r6
+             if w7 < 128 then return (val6 .|. (fromIntegral w7 `shiftL` 42), r7) else do
+              let val7 = (val6 .|. (fromIntegral (w7 .&. 0x7F) `shiftL` 42))
+
+              (w8,r8) <- eitherUncons r7
+              if w8 < 128 then return (val7 .|. (fromIntegral w8 `shiftL` 49), r8) else do
+               let val8 = (val7 .|. (fromIntegral (w8 .&. 0x7F) `shiftL` 49))
+
+               (w9,r9) <- eitherUncons r8
+               if w9 < 128 then return (val8 .|. (fromIntegral w9 `shiftL` 56), r9) else do
+                let val9 = (val8 .|. (fromIntegral (w9 .&. 0x7F) `shiftL` 56))
+
+                (w10,r10) <- eitherUncons r9
+                if w10 < 128 then return (val9 .|. (fromIntegral w10 `shiftL` 63), r10) else do
+
+                 Left ("failed to parse varint128: too big; " ++ show val6)
```

Comment:
It would be good to have a round-trip encode-decode QuickCheck test on random Word64 values.

Reply:
we have this via the uint64 tests (among others) -- there's just not much coverage. I beefed it up to do 10,000 in this case to stress test a bit that the higher encodings also aren't screwy.

Reply:
OK, thanks.